### PR TITLE
[Ready] Pvp gamemode

### PR DESCRIPTION
**Pull request by Kmc2000**
An experimental gamemode pitting players against each other.

-The Syndicate get a ship and must destroy Nanotrasen's finest by any means necessary
-The Syndicate get boarding equipment and a nuke. Load the nuke into a raptor to get it on station.
-The Syndicate can choose to weaken NT's ship in open combat with their syndicate reinforcements, or try and stealth it with raptors.
-Syndicate must destroy the ship within 45 minutes, or else they lose.

Everything will break. There will be fire. Hopefully it'll be fun.

## Changelog
:cl: Kmc2000 with Cenrus
add: Added a PVP gamemode (EXPERIMENTAL FOR NOW)
/:cl:

### CODE REVIEW

**Comment by Cenrus** [comment]
What did you change in regards to the Enterprise?

**Comment by Kmc2000** [comment]
> What did you change in regards to the Enterprise?

it was a mistake

like everything about this PR

**Comment by Cenrus** [comment]
We can only hope that it's a happy little accident.

File: nsv13/code/game/gamemodes/pvp/pvp.dm [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+	recommended_enemies = 10
+	antag_flag = ROLE_SYNDI_CREW
+	enemy_minimum_age = 0
+
+	announce_span = "danger"
+	announce_text = "Nanotrasen's incursions into Syndicate space have not gone unnoticed!\n\
+	<span class='danger'>Operatives</span>: Eliminate Nanotrasen's presence in Syndicate space using the weaponry of the SSV Hammurabi, or via a thermonuclear detonation.\n\
+	<span class='notice'>Crew</span>: Survive the Syndicate assault long enough for reinforcements to arrive."
+
+	title_icon = "nukeops"
```
Comment (Jalleo) [inline nsv13/code/game/gamemodes/pvp/pvp.dm]: Needs a actual icon

File: nsv13/code/game/gamemodes/pvp/pvp.dm [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+
+	var/list/pre_nukeops = list()
```
Comment (Jalleo) [inline nsv13/code/game/gamemodes/pvp/pvp.dm]: copy pasta? BAD

File: nsv13/code/game/gamemodes/pvp/pvp.dm [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+
+	var/nukes_left = 1
+
+	var/datum/team/nuclear/nuke_team
+	var/highpop_threshold = 45 //At what player count does the round enter "highpop" mode, and spawn the Syndicate a larger ship to compensate.
+
+	var/operative_antag_datum_type = /datum/antagonist/nukeop/syndi_crew
+	var/leader_antag_datum_type = /datum/antagonist/nukeop/leader/syndi_crew
+	var/time_limit
+	var/list/standard_ships = list("Hammurabi.dmm") //Update this list if you make more PVP ships :) ~Kmc
```
Comment (Jalleo) [inline nsv13/code/game/gamemodes/pvp/pvp.dm]: CONFIG

File: nsv13/code/game/gamemodes/pvp/pvp.dm [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+	var/list/highpop_ships = list("Hulk.dmm") //Update this list if you make a big PVP ship
```
Comment (Jalleo) [inline nsv13/code/game/gamemodes/pvp/pvp.dm]: DITTO

File: nsv13/code/game/gamemodes/pvp/pvp.dm [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+
+/datum/game_mode/pvp/pre_setup()
+	var/pop = num_players()
+	var/enemies_to_spawn = required_enemies + round((pop-required_enemies)/10) //Syndicates scale with pop. On a standard 30 pop, this'll be 30 - 10 -> 20 / 10 -> 2 floored = 2, where FLOOR rounds the number to a whole number.
+	var/map_file = pop < highpop_threshold ? pick(standard_ships) : pick(highpop_ships) //Scale the ship map to suit player pop. Larger crews need more space.
+	var/ship_type = pop < highpop_threshold ? /obj/structure/overmap/syndicate/pvp : /obj/structure/overmap/syndicate/pvp/hulk
+	var/obj/structure/overmap/syndiship
+	for(var/z in SSmapping.levels_by_trait(ZTRAIT_CORVI))
+		syndiship = new ship_type(get_turf(locate(round(world.maxx * 0.5, 1), round(world.maxy * 0.5, 1), z))) //Make a new syndie ship object in Corvi.
+	message_admins("Spawning in syndi ship map, this may take a while. No the game hasn't crashed, I'm just loading a map before we start.") //Warn the admins. This shit takes a while.
```
Comment (Jalleo) [inline nsv13/code/game/gamemodes/pvp/pvp.dm]: Can we instead have it so the PVP round is only available via specific map vote at the round end instead?

File: nsv13/code/game/gamemodes/pvp/pvp.dm [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+			log_game("[key_name(new_op)] has been selected as a syndicate crewmember")
+		addtimer(CALLBACK(src, .proc/force_lighting, syndiship), 6 SECONDS)
+		return TRUE
+	else
+		qdel(syndiship)
+		setup_error = "Not enough syndicate crew candidates"
+		return FALSE
+////////////////////////////////////////////////////////////////////////////////////////
+////////////////////////////////////////////////////////////////////////////////////////
+/datum/game_mode/pvp/proc/force_lighting(obj/structure/overmap/hammurabi)
```
Comment (Jalleo) [inline nsv13/code/game/gamemodes/pvp/pvp.dm]: LOAD THE MAP AS A MAP VOTE INSTEAD.

File: nsv13/code/game/gamemodes/pvp/pvp.dm [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+	return ..()
+
+/datum/game_mode/pvp/OnNukeExplosion(off_station)
+	..()
+	nukes_left--
+
+/datum/game_mode/pvp/check_win()
+	if (nukes_left == 0)
+		return TRUE
+	if(world.time >= time_limit)
```
Comment (Jalleo) [inline nsv13/code/game/gamemodes/pvp/pvp.dm]: Why are you checking this here is there is a timer that calls this after 45.5 minuites?

File: nsv13/code/modules/munitions/ship_weapons/_ship_weapon.dm [hunk 8]
```diff
@@ -94,6 +94,15 @@
  */
 /obj/machinery/ship_weapon/Initialize()
 	. = ..()
+	addtimer(CALLBACK(src, .proc/PostInitialize), 5 SECONDS)
```
Comment (Jalleo) [inline nsv13/code/modules/munitions/ship_weapons/_ship_weapon.dm]: We can just make the proc here be the damn get_ship() after five seconds.

File: nsv13/code/game/gamemodes/pvp/pvp.dm [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+				return FALSE
+	return ..()
+
+/datum/game_mode/pvp/set_round_result()
+	..()
+	var result = nuke_team.get_result()
+	if(result == NUKE_RESULT_NUKE_WIN)
+		SSticker.mode_result = "win - syndicate nuke"
+		SSticker.news_report = STATION_NUKED
+	else
```
Comment (Jalleo) [inline nsv13/code/game/gamemodes/pvp/pvp.dm]: Can the NT crew nuke the syndi ship?

File: nsv13/code/modules/overmap/types.dm [hunk 10]
```diff
@@ -161,6 +165,78 @@
...
+	damage_states = TRUE
+	area_type = /area/hammurabi
+	max_integrity = 1200 //Buffed health due to ship internal damage existing
+	integrity_failure = 1200
+	pixel_z = -32
+	pixel_w = -32
+	ai_controlled = FALSE
+
+
+/obj/structure/overmap/syndicate/pvp/hulk //Larger PVP ship for larger pops.
```
Comment (Cenrus) [inline nsv13/code/modules/overmap/types.dm]: Set a name for the Hulk so it's easy to distinguish

File: nsv13/code/controllers/subsystem/starsystem.dm [hunk 11]
```diff
@@ -31,6 +33,17 @@ SUBSYSTEM_DEF(starsystem)
 /datum/controller/subsystem/starsystem/New()
 	. = ..()
 	instantiate_systems()
+	for(var/type in subtypesof(/obj/structure/overmap/syndicate/ai))
+		if(!(locate(type) in enemy_blacklist))
+			enemy_types += type
```
Comment (covertcorvid) [inline nsv13/code/controllers/subsystem/starsystem.dm]:
```suggestion
	enemy_types = subtypesof(/obj/structure/overmap/syndicate/ai)
	for(var/type in enemy_blacklist)
		enemy_types -= type

```

File: nsv13/code/controllers/subsystem/starsystem.dm [hunk 12]
```diff
@@ -31,6 +33,17 @@ SUBSYSTEM_DEF(starsystem)
...
+		if(!(locate(type) in enemy_blacklist))
+			enemy_types += type
+
+
+/datum/controller/subsystem/starsystem/proc/add_blacklist(what)
+	enemy_types = list()
+	enemy_blacklist += what
+	for(var/type in subtypesof(/obj/structure/overmap/syndicate/ai))
+		if(!(locate(type) in enemy_blacklist))
+			enemy_types += type
```
Comment (covertcorvid) [inline nsv13/code/controllers/subsystem/starsystem.dm]:
```suggestion
	enemy_blacklist += what
	if(locate(what) in enemy_types)
			enemy_types -= what

```

File: nsv13/code/game/gamemodes/pvp/pvp.dm [hunk 13]
```diff
@@ -0,0 +1,237 @@
...
+	var/pop = num_players()
+	var/enemies_to_spawn = required_enemies + round((pop-required_enemies)/10) //Syndicates scale with pop. On a standard 30 pop, this'll be 30 - 10 -> 20 / 10 -> 2 floored = 2, where FLOOR rounds the number to a whole number.
+	var/map_file = pop < highpop_threshold ? pick(standard_ships) : pick(highpop_ships) //Scale the ship map to suit player pop. Larger crews need more space.
+	var/ship_type = pop < highpop_threshold ? /obj/structure/overmap/syndicate/pvp : /obj/structure/overmap/syndicate/pvp/hulk
+	var/obj/structure/overmap/syndiship
+	for(var/z in SSmapping.levels_by_trait(ZTRAIT_CORVI))
+		syndiship = new ship_type(get_turf(locate(round(world.maxx * 0.5, 1), round(world.maxy * 0.5, 1), z))) //Make a new syndie ship object in Corvi.
+	message_admins("Spawning in syndi ship map, this may take a while. No the game hasn't crashed, I'm just loading a map before we start.") //Warn the admins. This shit takes a while.
+	var/list/errorList = list()
+	var/list/loaded = SSmapping.LoadGroup(errorList, "Syndicate ship", "map_files/PVP", map_file, default_traits = ZTRAITS_MINING_SHIP, silent = FALSE)
```
Comment (covertcorvid) [inline nsv13/code/game/gamemodes/pvp/pvp.dm]: If the event isn't disabled, won't making it the "mining ship" occasionally spawn belt rats on top of them?

File: nsv13/code/modules/munitions/ship_weapons/_ship_weapon.dm (hunk 8 above)
Comment (covertcorvid) [reply nsv13/code/modules/munitions/ship_weapons/_ship_weapon.dm]: Also needs to set time to maintenance

File: nsv13/code/modules/power/stormdrive.dm [hunk 14]
```diff
@@ -58,7 +58,7 @@ Takes  plasma and outputs superheated plasma and a shitload of radiation.
 	anchored = TRUE
 	resistance_flags = INDESTRUCTIBLE | LAVA_PROOF | FIRE_PROOF | UNACIDABLE | ACID_PROOF | FREEZE_PROOF
 	light_color = LIGHT_COLOR_CYAN
-	var/obj/machinery/atmospherics/components/binary/pump/pipe
+	var/obj/machinery/atmospherics/components/binary/pump/reactor_manifold/pipe
```
Comment (covertcorvid) [inline nsv13/code/modules/power/stormdrive.dm]: This should really be a separate PR but whatever

File: nsv13/code/game/gamemodes/pvp/pvp.dm (hunk 13 above)
Comment (Kmc2000) [reply nsv13/code/game/gamemodes/pvp/pvp.dm]: The "Mining ship" ZTRAITS are just to allow it to be boardable, self looped space etc. We ought to actually refactor this to ZTRAITS_BOARADABLE_SHIP or such

Comment (Jalleo) [reply nsv13/code/game/gamemodes/pvp/pvp.dm]: Then that has to be done before this is merged.

**Comment by Vasily2013** [comment]
Map Change Requests

Hammurabi
- [ ] Create door to plasma tank in engineering
- [ ] add meters to the pipes in engineering.
- [ ] change pipes in engineering to visible
- [ ] Cell buttons overlap with one another.
- [ ] Make an atmospherics network
- [ ] Add air alarms
- [ ] add at least one (1) redundancy for power lines.
- [ ] add a space heater to the stormdrive core access room
- [ ] Clearly mark external access airlock, or the route to it, instead of hiding it in maint.
- [ ] Use syndicate sleepers you lazy bum.
- [ ] Remove two railguns for balance with Hulk so each ship has the same armaments

Hulk
- [ ] Change airlocks just north of the launch tubes to rapid cycle, they won't cycle link on their own in current orientation.
- [ ] Make an atmos network
- [ ] add air alarms
- [ ] if you're going to start a player ghost role in the Hulk then they should have some clever means of escape.
- [ ] Computer wall effects on bridge may be visible from outside.

**Comment by Vasily2013** [comment]
Idea: Two vagos running against one another.

**Comment by Cenrus** [comment]
Remove the Ghost spawner from the Hulk, I put it in there for a test, but forgot to remove it.

**Comment by Kmc2000** [comment]
Note to self:
Syndicate fighters need to emag on spawn, or they'll be unable to use ship weapons on the main ship.

File: nsv13/code/game/gamemodes/pvp/pvp.dm (hunk 3 above)
Comment (francinum) [reply nsv13/code/game/gamemodes/pvp/pvp.dm:39]: Feasible yet pointless at the moment.

File: nsv13/code/game/gamemodes/pvp/pvp.dm (hunk 4 above)
Comment (francinum) [reply nsv13/code/game/gamemodes/pvp/pvp.dm:40]: Feasible yet pointless at the moment.

File: nsv13/code/game/gamemodes/pvp/pvp.dm (hunk 5 above)
Comment (francinum) [reply nsv13/code/game/gamemodes/pvp/pvp.dm:60]: Against design goals.

File: nsv13/code/game/gamemodes/pvp/pvp.dm (hunk 9 above)
Comment (francinum) [reply nsv13/code/game/gamemodes/pvp/pvp.dm:159]: Valid question, @Kmc2000

File: nsv13/code/game/gamemodes/pvp/pvp.dm (hunk 7 above)
Comment (Kmc2000) [reply nsv13/code/game/gamemodes/pvp/pvp.dm:139]: Failsafe check. Yes there is a timer that checks this, but the nuke may have gone off before then.

File: nsv13/code/game/gamemodes/pvp/pvp.dm (hunk 2 above)
Comment (Kmc2000) [reply nsv13/code/game/gamemodes/pvp/pvp.dm:29]: Doesn't really affect anything

File: nsv13/code/game/gamemodes/pvp/pvp.dm (hunk 1 above)
Comment (Kmc2000) [reply nsv13/code/game/gamemodes/pvp/pvp.dm:27]: When I get time. For now, it'll do. Theyre functionally similar enough anyway

**Comment by Kmc2000** [comment]
finally AAAA
